src/audio/multiband_drc/multiband_drc.c

```diff
@@ -0,0 +1,605 @@
+			    "multiband_drc_new(): comp_data_blob_handler_new() failed.");
+		rfree(dev);
+		rfree(cd);
+		return NULL;
+	}
+
+	/* Get configuration data and reset DRC state */
+	ret = comp_init_data_blob(cd->model_handler, bs, ipc_multiband_drc->data);
+	if (ret < 0) {
+		comp_cl_err(&comp_multiband_drc,
+			    "multiband_drc_new(): comp_init_data_blob() failed.");
+		rfree(dev);
+		rfree(cd);
+		return NULL;
+	}
+	multiband_drc_reset_state(&cd->state);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void multiband_drc_free(struct comp_dev *dev)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "multiband_drc_free()");
+
+	comp_data_blob_handler_free(cd->model_handler);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static int multiband_drc_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_info(dev, "multiband_drc_params()");
+
+	comp_dbg(dev, "multiband_drc_verify_params()");
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "multiband_drc_params(): comp_verify_params() failed.");
+		return -EINVAL;
+	}
+
+	/* All configuration work is postponed to prepare(). */
+	return 0;
+}
+
+static int multiband_drc_cmd_get_data(struct comp_dev *dev,
+				      struct sof_ipc_ctrl_data *cdata,
+				      int max_size)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "multiband_drc_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
+		break;
+	default:
+		comp_err(dev, "multiband_drc_cmd_get_data() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int multiband_drc_cmd_set_data(struct comp_dev *dev,
+				      struct sof_ipc_ctrl_data *cdata)
+{
+	struct multiband_drc_comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "multiband_drc_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+		break;
+	default:
+		comp_err(dev, "multiband_drc_cmd_set_data() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int multiband_drc_cmd(struct comp_dev *dev, int cmd, void *data, int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "multiband_drc_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = multiband_drc_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = multiband_drc_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	default:
+		comp_err(dev, "multiband_drc_cmd(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int multiband_drc_trigger(struct comp_dev *dev, int cmd)
+{
+	int ret;
+
+	comp_info(dev, "multiband_drc_trigger()");
+
+	ret = comp_set_state(dev, cmd);
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
```

Comment:
where does this come from?

Reply:
COMP_STATUS_STATE_ALREADY_SET is returned when comp_set_state() requests the same state as dev's. In all component designs these codes are used in "prepare" and "trigger". Maybe these codes want to avoid the successive redundant call?
Hi @singalsu do you know about this?

Reply:
Usually the components just return the value from `comp_set_state(dev, cmd);`. In EQs the component copy function pointer seems to be checked with assert. Actually I don't understand why the constant code piece in prepare() to return PPL_STATUS_PATH_STOP is not done by the framework for normal "source -> process -> sink" components.

Reply:
@lgirdwood Do you have thoughts?

Reply:
@lgirdwood ping on this

Reply:
so this is valid since we don't do basic checks like this in the parent layer. I think this should not be the base case as this in my mind screams anti DRY coding. IMO we should do these checks in the base layer unless the module signals that we don't want them done. Same thing goes for handling component state in my opinion. In any case this is valid for this review.